Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractNodeStore.java

```diff
@@ -0,0 +1,527 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.tree;
+
+import static com.amazon.randomcutforest.CommonUtils.toFloatArray;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Stack;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import com.amazon.randomcutforest.MultiVisitor;
+import com.amazon.randomcutforest.Visitor;
+import com.amazon.randomcutforest.store.IPointStoreView;
+import com.amazon.randomcutforest.store.IntervalManager;
+
+/**
+ * A fixed-size buffer for storing interior tree nodes. An interior node is
+ * defined by its location in the tree (parent and child nodes), its random cut,
+ * and its bounding box. The NodeStore class uses arrays to store these field
+ * values for a collection of nodes. An index in the store can be used to look
+ * up the field values for a particular node.
+ *
+ * The internal nodes (handled by this store) corresponds to
+ * [0..upperRangeLimit]
+ *
+ * If we think of an array of Node objects as being row-oriented (where each row
+ * is a Node), then this class is analogous to a column-oriented database of
+ * Nodes.
+ *
+ * Note that a NodeStore does not store instances of the
+ * {@link com.amazon.randomcutforest.tree.Node} class.
+ */
+public abstract class AbstractNodeStore {
+
+    public static double SWITCH_FRACTION = 0.499;
+
+    protected final int capacity;
```

Comment:
I think this is either the number of internal nodes or the number of leaf nodes that we can store now. Should we change the name?

Reply:
yes.